azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
+
+                        # NOTE: There is a ~small~ chance of a false positive here if an unexpected
+                        # disconnection occurs while a ReauthorizationOperation is in flight.
+                        # However, it will sort itself out - the ensuing connect that ocurrs as part
+                        # of the reauthorization will restore connection (no harm done) of it will
+                        # fail, at which point the failure was a result of a manual operation and
+                        # reconnection is not supposed to occur. So either way, we end up where we want
+                        # to be despite the false positive - just be aware that this can happen.
+                        logger.debug(
+                            "{}({}): Not attempting to reconnect (Reauthorization in progress)".format(
+                                self.name, event.name
+                            )
+                        )
+
+                    # BAD STATE (this block should not be reached)
+                    else:
+                        logger.warning(
+                            "{}: DisconnectEvent received while in unexpected state - {}, Connected: {}".format(
+                                self.name, self.state, self.pipeline_root.connected
+                            )
+                        )
+
+            # In all cases the event is sent up
             self.send_event_up(event)
 
+        # Connection Retry disabled
         else:
             self.send_event_up(event)
 
     @pipeline_thread.runs_on_pipeline_thread
-    def _send_new_connect_op_down(self):
+    def _add_connection_op_callback(self, op):
+        """Adds callback to a connection op passing through to do necessary stage upkeep"""
         self_weakref = weakref.ref(self)
 
+        # NOTE: we are currently protected from connect failing due to being already connected
+        # by the ConnectionLockStage. If the ConnectionLockStage changes functionality,
+        # we may need some logic changes to address an op that can fail while leaving us CONNECTED
+
         @pipeline_thread.runs_on_pipeline_thread
-        def on_connect_complete(op, error):
+        def on_complete(op, error):
+            this = self_weakref()
+            # If error, set us back to a DISCONNECTED state. It doesn't matter what kind of
+            # connection op this was, any failure should result in a disconnected state.
+
+            # NOTE: Due to the stage waiting any ops if an ongoing connection op is in-progress
+            # as well as the way that the reconnection process checks if there is an in-progress
+            # connection op (and punts the reconnect if so), there is no risk here of setting
+            # directly to DISCONNECTED - the in-progress state being overwritten is always going
+            # to be due to this op that is now completing, we can be assured of that.
+            if error:
+                with self.state_lock:
```

Comment:
>self [](http://example.com/codeflow?start=21&length=4)

Using `self` in here defeats the use of the weak reference.   This should `with this.state_lock:`

Reply:
Fixed here and everywhere